File: flimpl.py
#! /usr/bin/env python

import sys
import os
import urllib
import shutil

class Flimpl:

	def run(self):
		try:
			func = getattr(self, sys.argv[1])
		# Command not found
		except AttributeError:
			print "The command", sys.argv[1], "doesn't exist"
		# No action defined, show help
		except IndexError:
			self.help()
		else:
			if callable(func):
				try:
					func()
				except IndexError:
					print "Missing argument for command '" + sys.argv[1] + "' check help"

	def help(self):
		print "Commands are as follows:"
		print "\tapp [name] - Creates files for new app."
		print "\tappdel [name] - Delete files from existing app."
		print "\tview [app] [name] - Create a view file for app."
		print "\treadme - Download README file"
		print "\tsample - Download Sample app."
		print "\thelp - Lists this information"

	def readme(self):
		# Prepare
		url = urllib.URLopener()

		# If the README file already exists, delete it
		if os.path.exists('README.markdown'):
			print "Readme already exists, deleting and downloding new.."
			os.remove('README.markdown')	

		# Download the file
		url.retrieve('http://github.com/Sirupsen/Flimpl-Extras/raw/master/README_FLIMPL.markdown', 'README.markdown')
		print "Done! Saved to ./README.markdown"

	def sample(self):
		# Prepare
		url = urllib.URLopener()

		# Download the controller for the sample
		print "Downloading controller.."
		url.retrieve('http://github.com/Sirupsen/Flimpl-Extras/raw/master/sample/controllers/sample.php', 'application/controllers/sample.php')
		print "Done!"

		# Then the model
		print "Downloading model.."
		url.retrieve('http://github.com/Sirupsen/Flimpl-Extras/raw/master/sample/models/sample.php', 'application/models/sample.php')
		print "Done!"
		
		# Finally, create the sample view folder and put the index into here
		print "Creating view folder and fetching sample view.."
		os.mkdir("application/views/sample")
		url.retrieve('http://github.com/Sirupsen/Flimpl-Extras/raw/master/sample/views/sample/index.php', 'application/views/sample/index.php')

		# All done!
		print "Done!"
	
	def view(self):
		# Create the view file
		index = open('application/views/' + sys.argv[2] + '/' + sys.argv[3] + '.php', 'wa')
		index.close()

	def app(self):
		# Open the controller for the new app
		controller = open('application/controllers/' + sys.argv[2] + '.php', 'wa')
		# Create the basic class
		controller.write('<?php\n\nclass ' + sys.argv[2] + '_Controller extends Controller {\n\n}')
		controller.close()

		# Open the model file and write the basic content
		model = open('application/models/' + sys.argv[2] + '.php', 'wa')
		model.write('<?php\n\nclass ' + sys.argv[2] + '_Model {\n\n}')
		model.close()

		# Make the view directory
		os.mkdir('application/views/' + sys.argv[2])

		# Create the view file for the index
		index = open('application/views/' + sys.argv[2] + '/index.php', 'wa')
		index.close()

		print "Files created for", sys.argv[2], "app"

	def appdel(self):
		# Confirm deletion
		input = raw_input("Sure you want to delete the " + sys.argv[2] + " app? ")

		# If any of the words in the tuple was the answer:
		if input in ('y', 'Y', 'yes'):
			# Check if the controller exists, then delete else give error
			if os.path.exists('application/controllers/' + sys.argv[2] + '.php'):
				os.remove('application/controllers/' + sys.argv[2] + '.php')
				print "Controller for", sys.argv[2], "app deleted"
			else:
				print "Controller for", sys.argv[2], "app not found"

			# Again, check if the model exists, then delete - else, error
			if os.path.exists('application/models/' + sys.argv[2] + '.php'):
				os.remove('application/models/' + sys.argv[2] + '.php')
				print "Model for", sys.argv[2], "app deleted"
			else:
				print "Model for", sys.argv[2], "app not found"
			
			# If the view directory for the app is present, delete it, else - error
			if os.path.exists('application/views/' + sys.argv[2]):
				shutil.rmtree('application/views/' + sys.argv[2])
				print "View for", sys.argv[2], "app deleted"
			else:
				print "View for", sys.argv[2], "app not found"
		# Negative to confirmation, stop program
		else:
			print "Not deleting app", sys.argv[2]

Flimpl().run()

File: flimpl.rb
#!/usr/bin/env ruby

require 'fileutils'
require 'open-uri'

# App list
class FlimplList < Hash
  # New instance
  def initialize(path)
    @path = path
 
    notallowed = ["..", "."]
 
    # Load existing apps
    Dir.foreach(@path + 'controllers/') do |file|
      unless notallowed.include?(file)
        add file.split(".")[0]
      end
    end
 
    self
  end
 
  # Add an app
  def add(app)
    if app.is_a?(App)
      self[app.name] = app
    else
      self[app] = App.new(app)
    end
  end
 
  # Write
  def write
    # For each of the apps
    each do |index, app|
      # Possible files and their contents
      @files = [
        ["controllers/#{app}.php", "<?php\n\nclass #{app}_Controller extends Controller {\n}"],
        ["models/#{app}.php", "<?php\n\nclass #{app}_Model {\n}"],
        ["views/#{app}"],
        ["views/#{app}/index.php"],
      ]

      new(app)
      delete(app)
      view(app)
    end
  end

  def view(app)
    app.view.each do |view|
      File.open(@path + "/views/#{app}/#{view}.php", 'w') {}
    end
  end

  def new(app)
    # If app wants to be ccreated
    if app.write?
      @files.each do |file|
        if File.exist?(@path + file[0])
          return false
        elsif file[0][-3..-1] == 'php'
          File.open(@path + file[0], 'w') { |f| f.write(file[1]) }
        else
          Dir.mkdir(@path + file[0])
        end
      end
    end
  end

  def delete(app)
    # If app wants to be deleted
    if app.delete?
      @files.each do |file|
        if File.directory?(@path + file[0])
          FileUtils.rm_rf(@path + file[0])
        elsif File.exist?(@path + file[0])
          File.delete(@path + file[0])
        end
      end
    end
  end
end
 
# App
class App
  attr_accessor :name, :write, :delete
  attr_reader :view
 
  alias_method :write?, :write
  alias_method :delete?, :delete
 
  def initialize(name, write = false, delete = false)
    @name = name
    @write = write
    @delete = delete
    @view = []
  end

  def view=(name)
    @view << name
  end

  def delete
    @delete = true
  end
 
  def to_s
    @name
  end
end

# Handling external content
class FlimplExternal
  # Download a file
  def download(file, url, input = false)
    if ["y", "Y"].include?(input) 
      File.delete(file)
    else 
      exit if input
    end

    raise "File #{file} already exist." if File.exist?(file)

    File.open(file, 'w') { |f| f.write(open(url).read) }
  end

  # Download sample files
  def sample(input = false)
    files = [
            ['application/controllers/sample.php', 'http://github.com/Sirupsen/Flimpl-Extras/raw/master/sample/controllers/sample.php'],
            ['application/models/sample.php', 'http://github.com/Sirupsen/Flimpl-Extras/raw/master/sample/models/sample.php'], 
            ['application/views/sample'],
            ['application/views/sample/index.php', 'http://github.com/Sirupsen/Flimpl-Extras/raw/master/sample/views/sample/index.php']
    ] 

    files.each do |file|
      if file[0][-3..-1] == 'php'
        download(file[0], file[1], input)
      else
        Dir.mkdir(file[0])
      end
    end
  end
end
 
# Command line interface
class CommandLineInterface
  # Instance
  def initialize
    @list = FlimplList.new("application/")	
    @ext = FlimplExternal.new

    # Handle command
    begin
      send ARGV[0]
      write
    rescue ArgumentError
      puts $!
    rescue TypeError
      help
    end
  end
 
  # List apps
  def list
    @list.each { |index, app| puts app.name.capitalize }
  end

  # Download sample
  def sample
    @ext.sample
  end

  # Create view
  def view
    raise ArgumentError, "Missing argument: [App name] for View" unless ARGV[1]
    raise ArgumentError, "Missing argument: [View name] for View" unless ARGV[2]

    begin
      @list[ARGV[1]].view = ARGV[2]
    # Couldn't call #view on object
    rescue NoMethodError
      puts "App. '" + ARGV[1] + "' doesn't exist."
    end
  end

  # Get readme
  def readme
    file = "README"
    url = "http://github.com/Sirupsen/Flimpl-Extras/raw/master/README_FLIMPL.markdown"

    begin
      @ext.download(file, url)
    rescue RuntimeError
      print file + " already exist. Confirm replacing: [Y/N] "
      input = $stdin.gets.strip
    end

    @ext.download(file, url, input)
  end

  def help
    puts <<-help
Commands for Flimpl.rb:
   app [name] - Creates files for new app.
   del [name] - Delete files from existing app.
   view [app] [name] - Create a view file for app.
   readme - Download README file
   sample - Download Sample app.
   help - Prints this information.
help
  end
 
  # Create app
  def app
    raise ArgumentError, "Missing argument: [App name] for View" unless ARGV[1]

    @list.add App.new(ARGV[1], true)
  end
 
  # Delete app
  def del
    raise ArgumentError, "Missing argument: [App name] for View" unless ARGV[1]

    ARGV[1..ARGV.length].each do |app|
      print "Confirm deletion of #{app}: [Y/N] "
      input = $stdin.gets.strip

      begin
        @list[app].delete = true if ["y", "Y"].include?(input)
      # App not found, #delete not found on object
      rescue NoMethodError
        puts "App '" + app + "' doesn't exist"
      end
    end
  end
 
  # Write
  def write
    @list.write
  end
end
 
CommandLineInterface.new
